Bitcoin syncing checks

rustyrussell opened the pull request:
This adds warning fields in getinfo, as well as preventing obviously-problematic operations when we're not synced.  We assume once we're synced, we stay synced, however.

Fixes: #2770

ZmnSCPxj commented on lightningd/peer_htlcs.c:
```diff
@@ -1442,6 +1442,20 @@ static bool peer_sending_revocation(struct channel *channel,
 }
 
+struct deferred_commigsig {
+	struct sync_waiter sync;
```
Wait what and then we `tal_free` via this field?

I tend to see this as bad style for C.  I know it is valid, but still bad style.

At least make a warning comment that *this* field has to be the first and that future refactorings should not move it from its position as first field, because we are going to `tal_free` via a pointer to this field.

rustyrussell commented on lightningd/chaintopology.h:
```diff
@@ -171,6 +182,15 @@ void begin_topology(struct chain_topology *topo);
+	return topo->sync_waiters == NULL;
+}
+
+/* This assumes waiter is a tal object: freeing it removes from waiter queue */
```
Yes, I considered this, but it means you need a double-allocation.  Perhaps it's worth it to maintain that pattern, however.

ZmnSCPxj replied:
Yes, synch is a one-time event and the extra allocs should not be significant.

rustyrussell commented on lightningd/chaintopology.c:
```diff
@@ -577,6 +589,17 @@ static void updates_complete(struct chain_topology *topo)
+				i->cb(i);
+				tal_free(i);
+			}
+			topo->sync_waiters = tal_free(topo->sync_waiters);
```
Yes, definitely.

ZmnSCPxj replied:
Probably deserves a comment in the header file?

rustyrussell commented:
OK, style is more standard now.

darosior commented:
`test_option_upfront_shutdown_script` is failing because we try to fund a channel just after starting. Maybe adding a `wait_for_log` with : "Bitcoind now synced" ?

rustyrussell commented on lightningd/chaintopology.c:
```diff
@@ -559,6 +559,31 @@ static void next_updatefee_timer(struct chain_topology *topo)
+
+static void destroy_sync_waiter(struct sync_waiter *waiter)
+{
+	list_del(&waiter->list);
```
No, it's not.  One line fix, to remove destructor in loop...

rustyrussell commented:
OK, with that one-liner, and a rebase for CHANGELOG.md, and a new commit to avoid spurious travis errors...

ZmnSCPxj commented:
CI seems to show that tests are flakier now, most of the flakes seem to be related to waiting for sync, maybe we have a race condition somewhere?

rustyrussell commented:
Yes.  bitcoind is a mess.  In regtest mode, it sometimes says it's still in ibd.  It's possible we should be ignoring this field, but I don't know even where to look; is it documented?  I'll try to work around it.

rustyrussell commented:
OK, two changes since last review:

1. We need to set ourselves to synced even if we don't add any blocks (immediate restart).
2. Made the tests less flaky by using an Event rather than sleep(10).

ZmnSCPxj commented:
ACK 4326c5b

cdecker commented on lightningd/bitcoind.c:
```diff
@@ -997,6 +997,103 @@ void bitcoind_getclientversion(struct bitcoind *bitcoind)
+		if (initial)
+			log_unusual(bitcoind->log,
+				    "Waiting for initial block download"
+				    " (this can take a while!)");
```
Won't really work since by that time we already have processed the current blockchain head - 14, meaning we have inserted a high-water-mark, and we'll restart scanning from that height. We might need a better heuristic for "not-yet-locked-in" start height allowing us to reset that high-water-mark on restart. But that's for another PR I'd say.

cdecker commented:
CI flakyness is resolved in a later PR.

ACK https://github.com/ElementsProject/lightning/pull/2912/commits/4326c5b064bef96f55fce5b40da3bbdcaf3faf0e
